dashboard/app/assets/javascripts/levels/contract_match.js.jsx

```diff
@@ -0,0 +1,257 @@
+    getInitialState: function () {
+      /**
+       * @param data {{
+       *   name: string,
+       *   rangeType: BlockValueType,
+       *   domainTypes: {
+       *     type: BlockValueType,
+       *     key: string
+       *   },
+       * }}
+       */
+      return {
+        name: "",
+        rangeType: blockValueType.NONE,
+        domainTypes: [
+          {
+            key: 'domain' + (this.maxDomainID++),
+            type: blockValueType.NONE,
+            order: 0
+          }
+        ]
+      }
+    },
+    onNameChangeEvent: function (event) {
+      this.setState({
+        name: event.target.value
+      });
+    },
+    onRangeChange: function (newType) {
+      this.setState({
+        rangeType: newType
+      })
+    },
+    onDomainChange: function (domainKey, newType) {
+      this.setState({
+        domainTypes:
+          $.map(this.state.domainTypes, function (object) {
+            if (object.key == domainKey) {
+              object.type = newType;
+            }
+            return object;
+          })
+      });
+    },
+    onDomainAdd: function () {
+      var nextDomainID = this.maxDomainID++;
+      this.setState({
+        domainTypes:
+          this.state.domainTypes.concat({
+            key: 'domain' + nextDomainID,
+            type: blockValueType.NONE,
+            order: nextDomainID
+          })
+      });
+    },
+    onDomainRemove: function (domainKey) {
+      this.setState({
+        domainTypes:
+          $.grep(this.state.domainTypes, function (object) {
+            return object.key !== domainKey;
+          })
+      });
+    },
+    render: function () {
+      return (
+        <div>
+          <div id='sectionTitle'>Name</div>
+          <div>
+            <input id='functionNameText' onChange={this.onNameChangeEvent} placeholder='Name' type='text' value={this.state.name}/>
+          </div>
+          <div id='sectionTitle'>Domain</div>
+          <DomainsList
+            domainTypes={this.state.domainTypes}
+            onDomainChange={this.onDomainChange}
+            onDomainAdd={this.onDomainAdd}
+            onDomainRemove={this.onDomainRemove}/>
+          <div id='sectionTitle'>Range</div>
+          <TypeChooser type={this.state.rangeType} onTypeChange={this.onRangeChange}/>
+        </div>
+      )
+    }
+  });
+
+  var DomainsList = React.createClass({
+    render: function () {
+      var self = this;
+      var sortedDomains = this.props.domainTypes.sort(function (a,b) {
+        return a.order > b.order;
+      });
+      var lastNode = this.props.domainTypes[this.props.domainTypes.length - 1];
+      var typeChoiceNodes = $.map(sortedDomains, function (object) {
+        var isLastNode = (object === lastNode);
+        return (
+          <div style={isLastNode ? {float: 'left'} : {}}>
+            <TypeChooser order={object.order} type={object.type} key={object.key}
+              onTypeChange={curry(self.props.onDomainChange, object.key)}/>
```

Comment:
When writing JSX for my personal code, I've found splitting each property onto its own line to help with readability.